File: TouchIDDemo_ObjC/TouchIDDemo_ObjC/ViewController.m
//
//  ViewController.m
//  TouchIDDemo_ObjC
//
//  Created by Pranay on 26/07/16.
//  Copyright © 2016 mobitronics. All rights reserved.
//

#import "ViewController.h"
#import <LocalAuthentication/LocalAuthentication.h>

#define kDefaultUsername    @"Username"
#define kDefaultPassword    @"Password"

@interface ViewController ()

@property (weak, nonatomic) IBOutlet UITextField *usernameTextField;
@property (weak, nonatomic) IBOutlet UITextField *passwordTextField;

@end

@implementation ViewController

- (void)viewDidLoad {
  [super viewDidLoad];

  UITapGestureRecognizer *tapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapGestureRecognizer:)];
  [self.view addGestureRecognizer:tapRecognizer];
}

- (void)didReceiveMemoryWarning {
  [super didReceiveMemoryWarning];
}

#pragma mark - TapGesture Method

- (void)tapGestureRecognizer:(UITapGestureRecognizer*)sender {
  //Hide Keyboard
  [self.usernameTextField resignFirstResponder];
  [self.passwordTextField resignFirstResponder];
}

#pragma mark - IBAction Method

- (IBAction)loginButtonAction:(id)sender {
  
  if (self.usernameTextField.text.length<=0 || self.passwordTextField.text.length<=0) {
    //Show alert for empty username or pasword
    [self showAlertMessage:@"username/password is empty!"];
  } else if (![self.usernameTextField.text isEqualToString:kDefaultUsername]) {
    //Show alert for wrong username
    [self showAlertMessage:@"username is not correct"];
  } else if (![self.passwordTextField.text isEqualToString:kDefaultPassword]) {
    //Show alert for wrong password
    [self showAlertMessage:@"password is not correct"];
  } else {
    [self.usernameTextField resignFirstResponder];
    [self.passwordTextField resignFirstResponder];
    //Validation Complete
    [self showAlertMessage:@"Logged in successfully!"];
  }
}

- (IBAction)authenticateButtonAction:(id)sender {
  
  LAContext *context = [[LAContext alloc] init];
  
  NSError *error = nil;
  if ([context canEvaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics error:&error]) {
    
    [context evaluatePolicy:LAPolicyDeviceOwnerAuthenticationWithBiometrics
            localizedReason:@"Logging in with Touch ID"
                      reply:^(BOOL success, NSError *error) {
                        
                        dispatch_async(dispatch_get_main_queue(), ^{
                          if (error) {
                            switch (error.code) {
                              case LAErrorAuthenticationFailed:
                                //Show Error Alert
                                [self showAlertMessage:@"There was a problem verifying your identity."];
                                break;
                              case LAErrorUserCancel:
                                //Show Error Alert
                                [self showAlertMessage:@"You pressed cancel."];
                                break;
                              case LAErrorUserFallback:
                                //Show Error Alert
                                [self showAlertMessage:@"You pressed password."];
                                break;
                                
                              default:
                                //Show Error Alert
                                [self showAlertMessage:@"Touch ID may not be configured"];
                                break;
                            }
                            
                            return;
                          }
                          
                          if (success) {
                            [self showAlertMessage:@"You are the device owner!"];
                          } else {
                            [self showAlertMessage:@"You are not the device owner."];
                          }
                        });
                      }];
    
  } else {
    //Show error alert message
    [self showAlertMessage:@"Your device cannot authenticate using TouchID."];
  }
}

#pragma mark - Other Methods

/**
 *  This method is used to show the alert message
 *
 *  @param message - message
 */
- (void)showAlertMessage:(NSString *)message {
  UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Alert" message:message preferredStyle:UIAlertControllerStyleAlert];
  [alertController addAction:[UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleDefault handler:nil]];
  [self presentViewController:alertController animated:YES completion:nil];
}

@end

File: TouchIDDemo_Swift/TouchIDDemo_Swift/ViewController.swift
//
//  ViewController.swift
//  TouchIDDemo_Swift
//
//  Created by Pranay on 26/07/16.
//  Copyright © 2016 mobitronics. All rights reserved.
//

import UIKit
import LocalAuthentication

class ViewController: UIViewController {
  
  @IBOutlet weak var usernameTextField: UITextField!
  @IBOutlet weak var passwordTextField: UITextField!

  var context = LAContext()
  
  let defaultUsername = "Username"
  let defaultPassword = "Password"
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    //Add TapGesture to View to hide the Keyboard
    let tapGesture = UITapGestureRecognizer.init(target: self, action: #selector(tapGestureRecognizer(tapGestureRecognizer:)))
    self.view.addGestureRecognizer(tapGesture)
  }

  override func didReceiveMemoryWarning() {
    super.didReceiveMemoryWarning()
  }

  //MARK: Tap Gesture Method
  /*
   Tap Gesture Recognizer method to recignize the tap and hide keyboard
   */
  func tapGestureRecognizer(tapGestureRecognizer: UITapGestureRecognizer) {
    usernameTextField.resignFirstResponder()
    passwordTextField.resignFirstResponder()
  }

  //MARK: Action Methods
  /*
   Login button action method
   */
  @IBAction func loginButtonAction(_ sender: AnyObject) {
    if (usernameTextField.text == "" || passwordTextField.text == "") {
      //Show alert for empty username/password
      showAlertMessage(message: "username/password is empty!")
    } else if (usernameTextField.text != defaultUsername || passwordTextField.text != defaultPassword) {
      //Show alert for wrong username or password
      showAlertMessage(message: "username/password is not correct!")
    } else {
      usernameTextField.resignFirstResponder()
      passwordTextField.resignFirstResponder()
      //Show success alert
      showAlertMessage(message: "Logged in successfully!")
    }
  }
  
  /*
   Touch ID buton Action
   */
  @IBAction func touchIDButtonAction(_ sender:AnyObject) {

    if context.canEvaluatePolicy(LAPolicy.deviceOwnerAuthenticationWithBiometrics, error:nil) {

      context.evaluatePolicy(LAPolicy.deviceOwnerAuthenticationWithBiometrics,
                             localizedReason: "Logging in with Touch ID",
                             reply: { (success : Bool, error : NSError? ) -> Void in
                              
                              DispatchQueue.main.async {
                                if success {
                                  //Show success alert message
                                  self.showAlertMessage(message: "Logged in with Touch ID")
                                }
                                
                                if error != nil {
                                  
                                  switch(error!.code) {
                                  case LAError.authenticationFailed.rawValue:
                                    self.showAlertMessage(message: "There was a problem verifying your identity.")
                                    break;
                                  case LAError.userCancel.rawValue:
                                    self.showAlertMessage(message: "You pressed cancel.")
                                    break;
                                  case LAError.userFallback.rawValue:
                                    self.showAlertMessage(message: "You pressed password.")
                                    break;
                                  default:
                                    self.showAlertMessage(message: "Touch ID may not be configured")
                                    break;
                                  }
                                }
                              }
                              
      })
    } else {
      self.showAlertMessage(message: "Touch ID not available")
    }
  }
  
  //MARK: Other Methods
  /*
   Show alert message
   */
  func showAlertMessage(message: String) {
    let alert = UIAlertController.init(title: "Alert", message: message, preferredStyle: UIAlertControllerStyle.alert)
    alert.addAction(UIAlertAction.init(title: "Ok", style: UIAlertActionStyle.default, handler: nil))
    self.present(alert, animated: true, completion: nil)
  }
}

